File: week2/readability.c
#include <cs50.h>
#include <math.h>
#include <ctype.h>
#include <string.h>
#include <stdio.h>

// Coleman - Liau index
// index = 0.0588 * L - 0.296 * S - 15.8
// where L is the average number of letters per 100 words in the text, and S is the average number of sentences per 100 words in the text.

int main(void)
{
    // get input from user
    string text = get_string("Text: ");

    // initialize counters and text length
    // 'words' var starts at 1 to account for last word, which won't have a space or .?! to up the counter
    int letters = 0;
    int words = 1;
    int sentences = 0;
    int length = strlen(text);

    // check whether character is a letter, a space, or end a sentence
    // and increase corresponding counter
    for (int i = 0; i <= length; i++)
    {
        if (isalpha(text[i]))
        {
            letters += 1;
        }
        else if ((int)text[i] == 32)
        {
            words += 1;
        }
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences += 1;
        }
    }

    // Calculates Coleman-Liau index
    float L = ((float)letters / (float)words) * 100;
    float S = ((float)sentences / (float)words) * 100;

    int index = round(0.0588 * L - 0.296 * S - 15.8);

    // Prints grade level, with lower and upper bounds of below 1 and 16+
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {

        printf("Grade %i\n", index);
    }
}

File: week6/readability.py
from cs50 import get_string
from regex import match


def get_text():
    """Get text from the user."""
    text = get_string("Text: ")
    return text


def counter_variables(text):
    """ Initialize some counter variables and eventually write them to a
    dict."""
    counters = {}
    letters = 0
    words = 1
    sentences = 0

    # Iterate over every letter of the text to see if it's a letter, a space, or
    # the end of a sentence, and increase the corresponding counter.
    for char in text:
        m = match(r'\w', char)
        punct = match(r'[.!?]', char)
        if m:
            letters += 1
        elif char == " ":
            words += 1
        elif punct:
            sentences += 1

    # Write the results of the counts to the dictionary and return it.
    counters["letters"] = letters
    counters["words"] = words
    counters["sentences"] = sentences

    return counters


def coleman_liau(counters):
    """Calculate Colmean-Liau index of user's text."""

    # Store the previously obtained counts to some variables.
    letters = counters["letters"]
    words = counters["words"]
    sentences = counters["sentences"]
    L = letters / words * 100
    S = sentences / words * 100

    # Get grade level rounded to the nearest integer
    grade = round(0.0588 * L - 0.296 * S - 15.8)
    return grade


def print_grade(grade):
    """Print reading level of the text."""

    if grade >= 16:
        print(f"Grade {grade}")
    elif grade < 1:
        print("Before Grade 1.")
    else:
        print(f"Grade {grade}")


def main():
    """Print the reading level of a given text based on the Coleman-Liau
    index."""

    text = get_text()
    counters = counter_variables(text)
    grade = coleman_liau(counters)
    print_grade(grade)
    return 0


main()
